substratevm/DebugInfo.md

```diff
@@ -76,12 +80,83 @@ created during population of the cache.
 Note that in all the examples above the `DebugInfoSourceSearchPath`
 options are actually redundant. In the first case, the classpath
 entries for _apps/hello/classes_ and _apps/greeter/classes_ will be used
-to derive the default search roots `apps/hello/src` and
+to derive the default search roots _apps/hello/src_ and
 _apps/greeter/src_. In the second case, the classpath entries for
 _apps/target/hello.jar_ and _apps/target/greeter.jar_ will be used to
 derive the default search roots _apps/target/hello-sources.jar_ and
 _apps/target/greeter-sources.jar_.
 
+## Debugging with Isolates
+
+Note that it is currently recommended to disable use of Isolates by
+passing flag `-H:-UseIsolates` on the command line when debug info
+generation is enabled. Enabling of Isolates affects the way that oops
+(object references) are encoded. In turn that means the debug info
+generator has to provide gdb with information about how to translate
+an encoded oop to the address in memory where the object data is
+stored. This sometimes requires care when asking gdb to process
+encoded oops vs decoded raw addresses.
+
+When isolates are disabled oops are essentially raw addresses pointing
+directly at the object contents. This is the same whether the oop is
+stored in a static/instance field or has been loaded into a register.
+
+When an oop is stored in a static or instance field gdb knows the type
+of the value stored in the field and knows how to dereference it to
+locate the object contents. For example, assume we have a `Units` type
+that details the scale used for a blueprint drawing and that the
+`Units` instance has a `String` field called `print_name`. Assume also
+we have a static field `DEFAULT_UNIT` that holds the standard `Units`
+instance. The following command will print the name for the default
+units.
+
+```
+(gdb) print *com.acme.Blueprint::DEFAULT_UNIT->print_name
+```
+
+gdb knows the type of the oop stored in `Blueprint::DEFAULT_UNIT` and
+knows how to dereference it to locate the object field
+values. Likewise, it knows that the `print_name` field is a `String`
+it will translate the oop stored in that field to an address where the
+String contents are stored and it will print the values of the
+`String` instance fields one by one.
+
+If, say, an oop referring to the `print_name` String has been loaded
+into $rdx it is still possible to print it using a straightforward
+cast to the pointer type that gdb associates with oop references.
+
+```
+(gdb) print/x *('java.lang.String' *)$rdx
```

Comment:
Interesting ... this approach is different than what we came up in EE debuginfo. There the dwarf-type for `java.lang.String` is a `DW_TAG_pointer_type` that points to a `DW_TAG_structure_type`. This allows us to encode the different deref-semantics when compressed pointers are used into the dwarf-types (i.e. we have `java.lang.String` and a  `_z_.java.lang.String` in case a location points to a String via compressed refs).

How are you going to solve that problem with your approach?

Reply:
Ahh I just saw the special special access for compressed refs is specified as dwarf expression in the DW_TAG_structure_type in field `DW_AT_data_location`. That means you can use the same approach. E.g for ` _z_.java.lang.String` we have:
```
DW_TAG_pointer_type
  DW_AT_type                  <0x002b5dd2>
  DW_AT_byte_size             0x00000004
DW_TAG_structure_type
  DW_AT_name                  _z_.java.lang.String
  DW_AT_byte_size             0x00000018
  DW_AT_data_location         0x970effffffff000000001a12302928050033247e0022 
  DW_AT_sibling               <0x002b5e24>
  DW_TAG_member
    DW_AT_name                  __hub__
    DW_AT_type                  <0x002d64e0>
    DW_AT_data_member_location  DW_OP_plus_uconst 0
  DW_TAG_member
    DW_AT_name                  value
  ...
```
where `DW_AT_data_location` holds the `dwarf-expression`-program that specifies the compressed refs access.

Reply:
Btw, do you know a tool that can disassemble the contents of `DW_AT_data_location` fields?

Reply:
First off, to answer the second question, objdump should decode data_location expressions when you dump the info section.

    $ objdump --dwarf=info hello
    . . . 
     <0><b>: Abbrev Number: 1 (DW_TAG_compile_unit)
        <c>   DW_AT_language    : 11	(Java)
     . . .
     <1><52>: Abbrev Number: 7 (DW_TAG_structure_type)
        <53>   DW_AT_name        : (indirect string, offset: 0x23af3): _objhdr
        <57>   DW_AT_byte_size   : 12
        <58>   DW_AT_data_location: 4 byte block: 97 37 20 1a 	(DW_OP_push_object_address; DW_OP_lit7; DW_OP_not; DW_OP_and)
    . . .

Ok, now for the pointer interpretation problem. I have thought about trying something similar to your solution but but I think there may be a problem with it. It's a tad complex so I hope you'll bear with me while I try to explain it in detail.

The root of the problem is actually the somewhat odd way DWARF supports re-interpretation of addresses using the data_location property. It expects that property to be located in the definition of the structure being pointed to not in the definition of the pointer that references the structure. That means that if you want to be able to distinguish, say, narrow pointers to struct Foo from direct pointers to struct Foo you actually need to have two equivalent definitions of the underlying layout type e.g. Struct Foo and struct _z_Foo, differing only in the presence or absence of the data_location.  In your case it comes down to having two names for pointers to two anonymous but equivalent structs and using those named pointer types strategically to label the relevant data locations, method arguments etc.

Of course, a narrow pointer and a direct pointer both actually point at the same type of object. However, because both anonymous structs have the same named fields any path expression that works for one struct works exactly the same for the other. So, the fact that two different types may be used to identify the same memory layout is transparent to whoever is driving gdb.

Your solution also doubles the amount of of DWARF info needed to define both the structs and their data members but that's maybe not too big a deal.

Now, in my code the names are attached to the struct types, not the pointer types. So, the equivalent trick would be to have two structs with two different names, define anonymous pointer types for each struct and and then use them strategically in the same way as you do to type pointer data. One reaso for using that naming model is that my code is not using structs, it is using classes which need to have name. Also, classes with different names are different types. Well, structs with different names are also different types.  Does that matter? There are several points where I think it does.

Apart from the base compilation unit that declares primitive an header types all other compilation units map 1-1 to an instance class or interface. Info records for members, i.e. instance fields, methods and static fields are declared within the scope of the method's owning class. Declarations (strictly specifications)are embedded as children of the class_type DIE. The latter two types of declaration have accompanying method definitions (essentially the method address ranges) and static field definitions (essentially address locations) embedded as level 1 siblings of the class_type DIE.

Providing a second class definition to carry the data_location attribute needed, say, to allow a narrow pointer to be translated would appear to require doubling all those elements. A consequence of this is that there would be more than one method definition associated with an address range and more than one data definition associated with a static field address. That's a problem because it means more than one DIE is associated with a given address range. gdb expects DIEs to be presented in ascending address order with no overlap.

Note that it would not be possible to get round this by putting the declarations in each of the two classes and have each equivalent declaration reference reference the equivalent definition record. Declarations and definitions must be cross-linked 1-1 in each direction. 

Is there a way around this? Well, it might be possible to omit the method and static field declarations from the copy, only making it structurally equivalent to the direct pointer target. However, that means the two type are to present differently to the user (at the very least command ptype would produce different output).

Another option is to use structs and not embed method and static field details in classes. That's perhaps ok in the short term. However, in the longer term I think it would really be better if the debugger knew that Java objects were objects and had associated behaviours. So, I'd really like to avoid switching from classes to structs and decoupling the methods and static fields if I can avoid it. 

One thing I still have to try is to define a narrow class type as a vacuous wrapper type that inherits from the direct pointer class with only one extra attribute i.e. the data_location field. I'm hoping we could use this as the base target for a narrow pointer type. I did try this some months back and it failed but that was when I was grappling with the bugs in the data_location processing. I'll retry this solution now that they are fixed and see if that can be used to enable us to declare the two alternative pointer types.

Reply:
Wow, just to make this a bit more salient ...  I noticed that symbol 'java.util.Scanner'::linePattern is ambiguos because it idenitfies a static field and a method of class scanner.

    (gdb) info var java.util.Scanner::linePattern
    All variables matching regular expression "java.util.Scanner::linePattern":
    
    File java/util/Scanner.java:
	    java.util.regex.Pattern *java.util.Scanner::linePattern;
    (gdb)  p * 'java.util.Scanner'::linePattern
    $16 = {java.util.regex.Pattern *(void)} 0x5f87d0 <java.util.Scanner::linePattern(void)>
    
    ptype 'java.util.Scanner'
    type = class java.util.Scanner : public java.lang.Object {
      private:
        int position;
        . . .
        static java.util.regex.Pattern *linePattern;
        . . .
        static java.util.regex.Pattern * linePattern(void);
        . . .

However, I then tried this and it shows that gdb really does know a lot about methods:

    (gdb) p * 'java.util.Scanner'::linePattern()
    $17 = {
      <java.lang.Object> = {
        <_objhdr> = {
          hub = 0x952c60,
          idHash = 157063191
        }, <No data fields>}, 
      members of java.util.regex.Pattern:
      flags = 0,
      pattern = 0xa2c110,
      . . .

So, a bit more info about calling conventions and we might be able to call methods direct from gdb!

Reply:
> One thing I still have to try is to define a narrow class type as a vacuous wrapper type that inherits from the direct pointer class with only one extra attribute i.e. the data_location field. I'm hoping we could use this as the base target for a narrow pointer type. I did try this some months back and it failed but that was when I was grappling with the bugs in the data_location processing. I'll retry this solution now that they are fixed and see if that can be used to enable us to declare the two alternative pointer types.

It that does not work maybe folding the data_location field in by declaring the type of each location that uses a compressed ref as a `DW_TAG_union_type` consisting of the regular `DW_TAG_class_type` plus a `DW_TAG_class_type` that only contains the `DW_AT_data_location`.

Reply:
> objdump --dwarf=info hello

Thanks for the info!

Reply:
> Another option is to use structs and not embed method and static field details in classes. That's perhaps ok in the short term. However, in the longer term I think it would really be better if the debugger knew that Java objects were objects and had associated behaviours. So, I'd really like to avoid switching from classes to structs and decoupling the methods and static fields if I can avoid it.

Once you settled on the dwarf type presentation for Java types you should not change it anymore if possible. Users might get upset if the way to access data changes from one release to the next. Also native-image specific debugging plugins for IDEs might pop into existence that start to rely on stable patterns on how to access certain kind data structures when debugging native-images. I.e. we should be really sure that the chosen type representation will work with compressed refs.

Reply:
> So, a bit more info about calling conventions and we might be able to call methods direct from gdb!

Sounds plausible. At least for methods that only require primitive type arguments.